users/users.go

```diff
@@ -1,228 +1,440 @@
+        }
+        if event.Event == RejoinedGroupEventType && lastRejoin == nil {
+            lastRejoin = &event
+        }
+        if lastBan != nil && lastRejoin != nil {
+            break
+        }
+    }
+
+    // If there is a Rejoin after a Ban, reset RankUpdatedAt to the Rejoin date
+    if lastBan != nil && lastRejoin != nil && lastRejoin.CreatedAt.After(lastBan.CreatedAt) {
+        log.Infof("User %s has rejoined after ban. Resetting RankUpdatedAt to rejoin date: %s", user.GetName(), lastRejoin.CreatedAt.Format("2006-01-02"))
+        user.RankUpdatedAt = &lastRejoin.CreatedAt
+        if err := db.DBCon.Save(&user).Error; err != nil {
+            return err
+        }
+    }
+
+    // Ensure RankUpdatedAt exists
+    if user.RankUpdatedAt == nil {
+        log.Warnf("User %s has no RankUpdatedAt – cannot proceed with rank calculation.", user.GetName())
+        return nil
+    }
+
+    // Find the current rank
+    currentRank := GetRankByName(user.RankName)
+    if currentRank == nil {
+        log.Warnf("Unknown current rank '%s' for user %s. Defaulting to first rank.", user.RankName, user.GetName())
+        currentRank = &Ranks[0]
+    }
+
+    // Calculate effective days
+    var effectiveDays int
+
+    if lastBan != nil && (lastRejoin == nil || lastRejoin.CreatedAt.Before(lastBan.CreatedAt)) {
+        // User is banned and has not rejoined yet -> only MinDays are counted
+        log.Infof("User %s is currently banned. Using MinDays for effective days: %d", user.GetName(), currentRank.MinDays)
+        effectiveDays = currentRank.MinDays
+    } else {
+        // Normal case: user is active -> calculate days since RankUpdatedAt
+        effectiveDays = int(time.Since(*user.RankUpdatedAt).Hours() / 24)
+    }
+
+    log.Infof("Effective days for user %s: %d", user.GetName(), effectiveDays)
+
+    promoted := false
+
+    // Promote the user through ranks if enough days have passed
+    for {
+        nextRank := GetNextRank(currentRank)
+        if nextRank == nil {
+            log.Infof("User %s already has the highest rank '%s'", user.GetName(), user.RankName)
+            break
+        }
+
+        daysNeededForNextRank := nextRank.MinDays - currentRank.MinDays
+        log.Infof("Days needed to go from '%s' to '%s': %d days", currentRank.Name, nextRank.Name, daysNeededForNextRank)
+
+        if effectiveDays >= daysNeededForNextRank {
+            log.Infof("Promoting user %s from '%s' to '%s'", user.GetName(), currentRank.Name, nextRank.Name)
+            user.RankName = nextRank.Name
+            user.RankUpdatedAt = ptrTimeNow()
+            currentRank = nextRank
+            effectiveDays -= daysNeededForNextRank
+            promoted = true
+
+            log.Infof("After promotion, effective days remaining for user %s: %d", user.GetName(), effectiveDays)
+
+        } else {
+            break
+        }
+    }
+
+    // Save the user only if a promotion happened
+    if promoted {
```

Comment:
Why not save it on the spot instead of waiting and keeping a boolean of `promoted`?